src/main/java/com/google/cloudassets/discovery/Main.java

```diff
@@ -22,46 +24,62 @@
-                                                        + "FROM information_schema.tables AS t "
-                                                        + "WHERE t.TABLE_NAME like '%Assets'";
-    private static final String GET_PROJECTS_LIST_QUERY = "SELECT t.workspaceId, t.projectId "
-                                                        + "FROM Workspace_Project_Table AS t "
-                                                        + "WHERE t.isActive = True";
+    private static final String GET_PROJECTS_LIST_QUERY = "SELECT workspaceId, projectId "
+                                                        + "FROM Workspace_Project_Table "
+                                                        + "WHERE isActive = True";
+    private static final String GET_TABLES_LIST_QUERY = "SELECT table_name "
+                                                        + "FROM information_schema.tables "
+                                                        + "WHERE table_name like '%Assets'";
+    // The order by part is important as the main asset table should be created before any other asset table
+    private static final String GET_SUPPORTED_TABLES_QUERY = "SELECT DISTINCT assetTableName, isMainTable "
+                                                        + "FROM Asset_Tables_Config "
+                                                        + "WHERE assetTableName != 'forAllAssets' "
+                                                        + "ORDER BY isMainTable DESC";
+
+    private static List<String> existingTableNames;
+    private static List<String> newSupportedTableNames;
 
     private static final FluentLogger logger = FluentLogger.forEnclosingClass();
 
     private static Spanner spanner;
     private static DatabaseId db;
     private static DatabaseClient dbClient;
     private static ReadOnlyTransaction readFromDb;
-    private static List<String> existingTableNames;
 
     /**
      * This function initializes all of the mapping and updating of all of the assets for each
      * of the projects in the ProjectId.txt file.
      */
     public static void main(String[] args) {
         SpannerOptions options = SpannerOptions.newBuilder().setProjectId(SPANNER_PROJECT_ID).build();
-        spanner = options.getService();
 
-        // Create spanner DatabaseClient
+        // Create spanner DatabaseClient and initialize static fields
+        spanner = options.getService();
         db = DatabaseId.of(SPANNER_PROJECT_ID, SPANNER_INSTANCE_ID, SPANNER_DATABASE_ID);
         dbClient = spanner.getDatabaseClient(db);
         readFromDb = dbClient.readOnlyTransaction();
-        createTablesIfNotExist();
 
         // Use spanner DatabaseClient
-        updateAllProjectsAssets();
+        try {
+            maintainTables();
+            updateAllProjectsAssets();
+        } catch (TableCreationException exception) {
+            String errorMsg = "Encountered an error while running table maintenance functions.";
```

Comment:
I "swallowed" the exceptions here in order to make sure that the db and spanner are properly closed even if an error is encountered. Once one of there exceptions occurs then our tables aren't really updated so the only concern is missing this failure. Do you have a better way to do this?

Reply:
I changed this to throw the exceptions as we discussed. However, in the end I used 'finally' instead of try-with-resources for two reasons:
1. In order to use try-with-resources the resources must be declared inside the try block - but I prefer these resources to be global
2. In between the initialisation of the 'spanner' object and the 'readFromDb' object there are two additional objects that must be initialised but they both do not implement java.io.Closeable (as they indeed do not need to be closed) and therefore can not be places in the try statment
Hope this makes sense to you